### PR TITLE
m2m-oauth-server: Managing M2M OAuth Server Tokens

### PR DESCRIPTION
**Pull request by jkralik**
### Overview
This pull request introduces comprehensive functionalities for managing tokens in the M2M OAuth server, enhancing the server's ability to handle token creation, retrieval, blacklisting, and deletion.

### Key Features

**Token Creation:**
- Support for generating access tokens with specific claims (subject, audience, scopes, expiration time).
- Crucial for secure authentication in machine-to-machine (M2M) interactions.

**Token Retrieval:**
- Tokens can be retrieved based on various criteria: owner, token ID, and blacklisted status.
- Facilitates efficient management and monitoring of issued tokens.

**Token Blacklisting:**
- Ability to blacklist tokens, preventing further use.
- Enhances security by identifying and managing compromised or obsolete tokens.

**Token Deletion:**
- Automatic deletion of expired tokens from the store.
- Maintains a clean and up-to-date token database, ensuring optimal performance and security.

**Configuration Enhancements:**
- Updates to configuration structures for more flexible and secure management of OAuth clients and their configurations.

**Protocol Buffers and gRPC:**
- Introduction of Protocol Buffers definitions and gRPC services for token management operations.
- Facilitates efficient communication and integration with other services.
- Definitions include messages for creating, retrieving, and blacklisting tokens, along with HTTP and OpenAPI annotations for RESTful API generation.

**Documentation:**
- Detailed documentation in Markdown and HTML formats.
- Describes the protocol, message types, and service methods.
- Ensures developers can easily understand and integrate the new functionalities.

**Web interface:**
- Users are able to manage their tokens.

### Benefits
These enhancements make the M2M OAuth server more robust and secure, providing essential features for managing API access tokens in a machine-to-machine context.

### CODE REVIEW

File: m2m-oauth-server/oauthSigner/oauthSigner.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+func setKeyError(key string, err error) error {
+	return fmt.Errorf("failed to set %v: %w", key, err)
+}
+
+type OAuthSigner struct {
+	privateKeyJWTValidators map[string]*validator.Validator
+	closer                  fn.FuncList
+	Config                  Config
```
Comment (Danielius1922) [inline m2m-oauth-server/oauthSigner/oauthSigner.go]: Do we want to allow other packages to modify the Config? If not, the I would go for unexported config and a getter

File: m2m-oauth-server/pb/tags.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package pb
+
+const (
+	ExpirationKey           = "expiration"
+	OwnerKey                = "owner"
+	BlackListedFlagKey      = "blacklisted.flag"
```
Comment (Danielius1922) [inline m2m-oauth-server/pb/tags.go]: BlackListedKey + ".flag" ?

File: m2m-oauth-server/service/grpc/service.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("cannot create grpc server options: %w", err)
+	}
+	server, err := server.New(config, fileWatcher, logger, opts...)
+	if err != nil {
+		return nil, err
+	}
+	pb.RegisterM2MOAuthServiceServer(server.Server, m2mOAuthServiceServer)
+
+	// SnippetService needs to stop gracefully to ensure that all commands are processed.
```
Comment (Danielius1922) [inline m2m-oauth-server/service/grpc/service.go]: M2MOAuthService

**Comment by jkralik** [comment]
@coderabbitai generate a description for this pullrequest about the managing m2m-oauth-server tokens feature targeting to not plgd developers.

File: charts/plgd-hub/templates/m2m-oauth-server/config.yaml [hunk 4]
```diff
@@ -17,16 +17,49 @@ data:
...
+          permitWithoutStream: {{ .apis.grpc.enforcementPolicy.permitWithoutStream }}
+        keepAlive:
+          # 0s - means infinity
+          maxConnectionIdle: {{ .apis.grpc.keepAlive.maxConnectionIdle }}
+          # 0s - means infinity
+          maxConnectionAge: {{ .apis.grpc.keepAlive.maxConnectionIdle }}
+          # 0s - means infinity
+          maxConnectionAgeGrace: {{ .apis.grpc.keepAlive.maxConnectionAgeGrace }}
+          time: {{ .apis.grpc.keepAlive.maxConnectionIdle }}
+          timeout: {{ .apis.grpc.keepAlive.maxConnectionIdle }}
```
Comment (Danielius1922) [inline charts/plgd-hub/templates/m2m-oauth-server/config.yaml]: `maxConnectionAge`, `time` and `timeout` are all set to `{{ .apis.grpc.keepAlive.maxConnectionIdle }}` is that correct?

File: m2m-oauth-server/store/store.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+var (
+	ErrNotSupported    = errors.New("not supported")
+	ErrNotFound        = errors.New("not found")
+	ErrNotModified     = errors.New("not modified")
+	ErrInvalidArgument = errors.New("invalid argument")
+	ErrPartialDelete   = errors.New("some errors occurred while deleting")
+)
+
+func IsDuplicateKeyError(err error) bool {
```
Comment (Danielius1922) [inline m2m-oauth-server/store/store.go]: rm? doesn't seem to be used

File: m2m-oauth-server/swagger.yaml [hunk 6]
```diff
@@ -18,6 +31,12 @@ paths:
                 client_id:
                   type: string
                   description: "The client ID."
+                token_name:
+                  type: string
+                  description: "The name of the token which will be used in the name claim."
+                expiration:
+                  type: integer
+                  description: "The requested expiration time in seconds from the client. If not provided, the token will be max allowed by client."
```
Comment (Danielius1922) [inline m2m-oauth-server/swagger.yaml]: `the token will be max allowed by client` is missing something, maybe - the token will have the max expiration time allowed by the client?

File: m2m-oauth-server/service/grpc/token.go [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+func getExpirationTime(clientCfg *oauthsigner.Client, tokenReq tokenRequest) time.Time {
+	if clientCfg.AccessTokenLifetime == 0 {
+		if tokenReq.CreateTokenRequest.GetExpiration() > 0 {
+			return time.Unix(tokenReq.CreateTokenRequest.GetExpiration(), 0)
+		}
+		return time.Time{}
+	}
+	if tokenReq.CreateTokenRequest.GetExpiration() == 0 {
+		if clientCfg.AccessTokenLifetime == 0 {
```
Comment (Danielius1922) [inline m2m-oauth-server/service/grpc/token.go]: this will never execute since it is checked already on line 111 and there it will always return from the function

File: m2m-oauth-server/service/http/requestHandler.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+// RequestHandler for handling incoming request
+type RequestHandler struct {
+	config                *Config
+	m2mOAuthServiceServer *grpcService.M2MOAuthServiceServer
+	mux                   *runtime.ServeMux
+}
+
+// NewHTTP returns HTTP handler
```
Comment (Danielius1922) [inline m2m-oauth-server/service/http/requestHandler.go]: NewHTTP -> NewRequestHandler

File: m2m-oauth-server/store/mongodb/tokens_test.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+	"github.com/plgd-dev/hub/v2/m2m-oauth-server/test"
+	"github.com/plgd-dev/hub/v2/test/config"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetTokens(t *testing.T) {
+	s, cleanUpStore := test.NewMongoStore(t)
+	defer cleanUpStore()
+
+	ctx, cancel := context.WithTimeout(context.Background(), config.TEST_TIMEOUT*100)
```
Comment (Danielius1922) [inline m2m-oauth-server/store/mongodb/tokens_test.go]: config.TEST_TIMEOUT

File: m2m-oauth-server/service/grpc/token.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+	}
+	for _, e := range s {
+		delete(check, e)
+	}
+	return len(check) == 0
+}
+
+func validateExpiration(clientCfg *oauthsigner.Client, tokenReq *tokenRequest) error {
+	if tokenReq.CreateTokenRequest.GetExpiration() > 0 {
+		if tokenReq.CreateTokenRequest.GetExpiration() < tokenReq.issuedAt.Unix() {
```
Comment (Danielius1922) [inline m2m-oauth-server/service/grpc/token.go:181]: add test case with non-zero expiration so we get better code coverage

File: m2m-oauth-server/store/mongodb/tokens_test.go (hunk 9 above)
Comment (jkralik) [reply m2m-oauth-server/store/mongodb/tokens_test.go]: fixed

File: m2m-oauth-server/service/http/requestHandler.go (hunk 8 above)
Comment (jkralik) [reply m2m-oauth-server/service/http/requestHandler.go]: fixed

File: m2m-oauth-server/service/grpc/token.go (hunk 7 above)
Comment (jkralik) [reply m2m-oauth-server/service/grpc/token.go]: removed

File: m2m-oauth-server/store/store.go (hunk 5 above)
Comment (jkralik) [reply m2m-oauth-server/store/store.go]: removed

File: m2m-oauth-server/oauthSigner/oauthSigner.go (hunk 1 above)
Comment (jkralik) [reply m2m-oauth-server/oauthSigner/oauthSigner.go]: fixed

File: m2m-oauth-server/pb/tags.go (hunk 2 above)
Comment (jkralik) [reply m2m-oauth-server/pb/tags.go]: fixed

File: m2m-oauth-server/service/grpc/service.go (hunk 3 above)
Comment (jkralik) [reply m2m-oauth-server/service/grpc/service.go]: fixed

File: charts/plgd-hub/templates/m2m-oauth-server/config.yaml (hunk 4 above)
Comment (jkralik) [reply charts/plgd-hub/templates/m2m-oauth-server/config.yaml]: fixed

File: m2m-oauth-server/pb/token.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+	if x.GetClientId() == "" {
+		return errors.New("Token.ClientId is empty")
+	}
+	if x.GetIssuedAt() == 0 {
+		return errors.New("Token.Timestamp is empty")
+	}
+	return nil
+}
+
+func (x *Token) ToMap() (map[string]interface{}, error) {
```
Comment (Danielius1922) [inline m2m-oauth-server/pb/token.go]: The methods below are for encoding to BSON right? 

There are similar pkg methods `UnmarshalProtoBSON`, `MarshalProtoBSON` and `ConvertStringValueToInt64` that I created when I was implementing snippet-service. I think they do almost the same stuff, maybe we can combine them into a single encode/decode-bson API

File: m2m-oauth-server/swagger.yaml (hunk 6 above)
Comment (jkralik) [reply m2m-oauth-server/swagger.yaml]: fixed

File: m2m-oauth-server/service/service.go [hunk 12]
```diff
@@ -4,82 +4,152 @@ import (
...
+	}
+	if config.CleanUpDeletedTokens != "" {
+		scheduler, err2 := NewExpiredUpdatesChecker(config.CleanUpDeletedTokens, config.ExtendCronParserBySeconds, func() {
+			err2 := s.DeleteTokens(ctx, time.Now())
+			if err2 != nil {
+				log.Errorf("cannot delete expired tokens: %v", err2)
+			}
+		})
+		if err2 != nil {
+			return nil, fmt.Errorf("cannot create scheduler: %w", err2)
```
Comment (Danielius1922) [inline m2m-oauth-server/service/service.go:56]: s.Close() before returning?

File: m2m-oauth-server/service/grpc/token.go (hunk 10 above)
Comment (jkralik) [reply m2m-oauth-server/service/grpc/token.go:181]: added

File: m2m-oauth-server/service/service.go (hunk 12 above)
Comment (jkralik) [reply m2m-oauth-server/service/service.go:56]: fixed

File: m2m-oauth-server/pb/token.go (hunk 11 above)
Comment (jkralik) [reply m2m-oauth-server/pb/token.go]: implemented

File: m2m-oauth-server/store/mongodb/tokens.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+	} else {
+		setIdOwnerHint = false
+	}
+	if owner != "" {
+		filter = append(filter, bson.E{Key: pb.OwnerKey, Value: owner})
+	} else {
+		setIdOwnerHint = false
+	}
+	if len(req.GetAudienceFilter()) > 0 {
+		filter = append(filter, bson.E{Key: pb.AudienceKey, Value: bson.M{mongodb.In: req.GetAudienceFilter()}})
```
Comment (Danielius1922) [inline m2m-oauth-server/store/mongodb/tokens.go]: add test with audienceFilter (according to SC this code path is not tested)

File: pkg/mongodb/marshal.go [hunk 14]
```diff
@@ -2,68 +2,258 @@ package mongodb
...
+func findParents(current any, permitMissingPaths bool, parts []string) ([]any, error) {
+	for idx, part := range parts {
+		if part == "" {
+			continue
+		}
+		switch curr := current.(type) {
+		case map[string]any:
+			var err error
+			current, err = iterateOverMap(curr, permitMissingPaths, part)
+			if err != nil {
```
Comment (Danielius1922) [inline pkg/mongodb/marshal.go]:
```go
if err != nil {
  return nil, err
}
if current == nil {
  return nil, nil
}
```

These lines are the same for both cases, if we move them outside of the switch SC should stop complaining about function complexity

File: m2m-oauth-server/service/grpc/token.go [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+func getExpirationTime(clientCfg *oauthsigner.Client, tokenReq tokenRequest) time.Time {
+	if clientCfg.AccessTokenLifetime == 0 {
+		if tokenReq.CreateTokenRequest.GetExpiration() > 0 {
+			return time.Unix(tokenReq.CreateTokenRequest.GetExpiration(), 0)
+		}
+		return time.Time{}
+	}
+	if tokenReq.CreateTokenRequest.GetExpiration() == 0 {
```
Comment (Danielius1922) [inline m2m-oauth-server/service/grpc/token.go]: Can we add a test case that goes through lines 117-125? Otherwise we're fine, even if we don't hit 80%, the rest of the code paths are error paths

File: m2m-oauth-server/store/mongodb/tokens.go (hunk 13 above)
Comment (jkralik) [reply m2m-oauth-server/store/mongodb/tokens.go]: I will remove audience part for now

File: pkg/mongodb/marshal.go (hunk 14 above)
Comment (jkralik) [reply pkg/mongodb/marshal.go]: fixed

File: m2m-oauth-server/service/grpc/token.go (hunk 15 above)
Comment (jkralik) [reply m2m-oauth-server/service/grpc/token.go]: added

**Review by Danielius1922 — APPROVED** [review]
